storage/innobase/include/row0purge.h

```diff
@@ -89,155 +89,69 @@ struct trx_purge_rec_t
-				handle */
-	dtuple_t*	row;	/*!< NULL, or a copy (also fields copied to
-				heap) of the indexed fields of the row to
-				handle */
-	dict_index_t*	index;	/*!< NULL, or the next index whose record should
-				be handled */
-	mem_heap_t*	heap;	/*!< memory heap used as auxiliary storage for
-				row; this must be emptied after a successful
-				purge of a row */
-	ibool		found_clust;/*!< whether the clustered index record
-				determined by ref was found in the clustered
-				index, and we were able to position pcur on
-				it */
-	btr_pcur_t	pcur;	/*!< persistent cursor used in searching the
-				clustered index record */
-#ifdef UNIV_DEBUG
-	/** whether the operation is in progress */
-	bool		in_progress;
-#endif
-	trx_id_t	trx_id;	/*!< trx id for this purging record */
-
-	/** meta-data lock for the table name */
-	MDL_ticket*		mdl_ticket;
-
-	/** table id of the previous undo log record */
-	table_id_t		last_table_id;
-
-	/** purge thread */
-	THD*			purge_thd;
-
-	/** metadata lock holds for this number of undo log recs */
-	int			mdl_hold_recs;
+/** Purge worker context */
+struct purge_node_t
+{
+  /** node type: QUE_NODE_PURGE */
+  que_common_t common;
 
-	/** Undo recs to purge */
-	std::queue<trx_purge_rec_t>	undo_recs;
+  /** DB_TRX_ID of the undo log record */
+  trx_id_t trx_id;
+  /** DB_ROLL_PTR pointing to undo log record */
+  roll_ptr_t roll_ptr;
 
-	/** Constructor */
-	explicit purge_node_t(que_thr_t* parent) :
-		common(QUE_NODE_PURGE, parent),
-		unavailable_table_id(0),
-		table(NULL),
-		heap(mem_heap_create(256)),
+  /** undo number of the record */
+  undo_no_t undo_no;
+
+  /** record type: TRX_UNDO_INSERT_REC, ... */
+  byte rec_type;
+  /** compiler analysis info of an update */
+  byte cmpl_info;
+  /** whether the clustered index record determined by ref was found
+  in the clustered index of the table, and we were able to position
+  pcur on it */
+  bool found_clust;
 #ifdef UNIV_DEBUG
-		in_progress(false),
+  /** whether the operation is in progress */
+  bool in_progress= false;
 #endif
-		mdl_ticket(NULL),
-		last_table_id(0),
-		purge_thd(NULL),
-		mdl_hold_recs(0)
-	{
-	}
+  /** table where purge is done */
+  dict_table_t *table= nullptr;
+  /** update vector for a clustered index record */
+  upd_t *update;
+  /** row reference to the next row to handle, or nullptr */
+  const dtuple_t *ref;
+  /** nullptr, or a deep copy of the indexed fields of the row to handle */
+  dtuple_t *row;
+  /** nullptr, or the next index of table whose record should be handled */
+  dict_index_t *index;
+  /** memory heap used as auxiliary storage; must be emptied between rows */
+  mem_heap_t *heap;
+  /** persistent cursor to the clustered index record */
+  btr_pcur_t pcur;
+
+  /** Undo recs to purge */
+  std::queue<trx_purge_rec_t> undo_recs;
+
+  /** map of table identifiers to table handles and meta-data locks */
+  std::unordered_map<table_id_t, std::pair<dict_table_t*,MDL_ticket*>> tables;
```

Comment:
Is there any sense to point buckets count explicitly here? We don't know what exactly value stl proposes for us by default.

Reply:
I checked that `libstdc++` 6.0.32 would default to `_M_h._M_bucket_count=1` and supposedly resize and rehash on demand. e4e00a07bda3ceb0b2dc569f9cb13860b1c0bafa introduces `constexpr size_t TRX_PURGE_TABLE_BUCKETS=128`, which I hope will be a reasonable compromise.